File: trader/macd.go
package trader

// Macd moving average convergence divergence
type Macd struct {
	EmaShort *ExponentialMovingAverage
	EmaLong  *ExponentialMovingAverage
	Current  float64
	Signal   string
}

// NewMacd constructor
func NewMacd(short int64, long int64, initial float64) *Macd {
	macd := &Macd{}
	macd.EmaShort = NewEma(short, initial)
	macd.EmaLong = NewEma(long, initial)
	macd.Current = 0
	macd.Signal = "wait"
	return macd
}

// Update updates macd and internal ema
func (macd *Macd) Update(closing float64) {
	macd.EmaShort.Update(closing)
	macd.EmaLong.Update(closing)
	before := macd.Current
	macd.Current = macd.EmaShort.Current - macd.EmaLong.Current
	if before < 0 && macd.Current > 0 {
		macd.Signal = "buy"
	} else if before > 0 && macd.Current < 0 {
		macd.Signal = "sell"
	} else {
		macd.Signal = "wait"
	}
}

File: python/macd.py
from ema import MovingAverage

class ConvergeDiverge:
    
    def __init__(self, short, long, initial):
        self.short = MovingAverage(short, initial)
        self.long = MovingAverage(long, initial)
        self.current = 0
        self.signal = 'wait'
        
        
    def update(self, value):
        self.short.update(value)
        self.long.update(value)
        before = self.current
        self.current = self.short.current - self.long.current
        if before < 0 and self.current > 0:
            self.signal = "buy"
        elif before > 0 and self.current < 0:
            self.signal = "sell"
        else:
            self.signal = "wait"